### ISSUE TITLE
getCpuid() causes crash (v0.60.0 git Release build, macOS 10.11.6)

### ISSUE THREAD

**Issue by zelbrium** [etternagame/etterna#321]
`getCpuid()`, from `StageStats.cpp`, (which uses inline assembly) appears to be causing some kind of stack corruption on Release builds, resulting in crashing errors within the next few functions (usually in `getSystemUniqueId()` when trying to modify `str`, but also sometimes in `getCpuHash()`). This is indirectly called at the end of a song, so that's not very nice.

I am building from git tag v0.60.0, macOS 10.11.6. On Debug builds, or when debugging with Xcode at all, the crash does not always occur. This bug may be related to currently-open #274, since it has the same error message, but it's a pretty generic error message, so maybe not.

**To fix**
I was able to fix the crash by using `__cpuid()` or `__get_cpuid()` from `cpuid.h` (same as what the Windows version of `getCpuHash()` uses) instead of the custom-built function `getCpuid()`. On my machine at least, both the custom function and the ones from `cpuid.h` return the same values.

I am not familiar enough with inline assembly calls to figure out why `getCpuid()` is causing errors.

**Comment by nico-abram** [comment]
Does it work if you do
```
#pragma optimize( "", off )
static void __attribute__ ((optnone)) 
getCpuid(uint32_t* p, uint32_t ax) 
{
	__asm __volatile("movl %%ebx, %%esi\n\t"
					 "cpuid\n\t"
					 "xchgl %%ebx, %%esi"
					 : "=a"(p[0]), "=S"(p[1]), "=c"(p[2]), "=d"(p[3])
					 : "0"(ax));
}
#pragma optimize( "", on )
```

**Comment by zelbrium** [comment]
No, however, using that method to disable optimizations on `getCpuHash()` instead, which calls `getCpuid()`, DOES work. I assume that has something to do with `getCpuid()` being inlined and so the optimization flags need to be applied to the calling function instead of the function itself. However, I had already tried disabling inlining on functions related to the issue and it didn't have an effect. Strange.

Anyway, disabling optimizations on `getCpuHash()` appears to be a solution. Maybe do it on both functions to be safe.

Just curious, is there a reason `_cpuid()` can't be used here? I assume it is not available on all the necessary platforms?

**Comment by nico-abram** [comment]
Not particularly. If we can get someone to test wether or not it works on linux with a direct call to that we can just use that. You can submit a pull request if you want and know how to

**Comment by zelbrium** [comment]
I'll try to test it on Linux within the next few days and post what I find here. Thanks!

**Comment by nico-abram** [comment]
theropfather is testing on linux right now, I also made a branch with the change so travis C.I tries to compile it on both mac and linux (https://travis-ci.org/etternagame/etterna/builds/438418913?utm_source=github_status&utm_medium=notification). I'll get back to you soon.

Just in case, this is the change you meant right? https://github.com/etternagame/etterna/commit/43ce3dfa91d7aea6d734f552bfa52fff492720e1

Also, thank YOU for finding and basically solving this :)

**Comment by zelbrium** [comment]
The syntax is a bit different from the Windows version to the Linux/mac version, it seems. It should be:

`__cpuid(0, cpuinfo[0], cpuinfo[1], cpuinfo[2], cpuinfo[3]);`

Alternatively, there is

`__get_cpuid(0, &cpuinfo[0], &cpuinfo[1], &cpuinfo[2], &cpuinfo[3]);`

which does the same thing but Wikipedia claims is safer ("It checks for extended features and does some more safety checks."), I guess.

**Comment by TheROPFather** [comment]
OK, should be fixed by 09d88d68fbe91131122ec7c88760a22c776b79d2. Will be in the next release, which is probably today so if you can test it then or compile, that'd be cool :+1:

**Comment by zelbrium** [comment]
@TheROPFather In that commit, 09d88d68fbe91131122ec7c88760a22c776b79d2, you wrote `__cpuid(cpuinfo, 0);`, which is not valid (at least, not on my Mac!). See my above message on the correct usage.

**Comment by TheROPFather** [comment]
OK, Ill try that out, thanks for the work.

**Comment by zelbrium** [comment]
It's the other way around according to my cpuid.h (__get_cpuid calls the __cpuid macro). But yeah, they're the same, so either one should work.

**Comment by TheROPFather** [comment]
Should actually be fixed this time 4378f4cca6bd737540a2ec16a21ec2ecb3eaa85d. Thank you for your input and work, you made this very easy for us :)
